src/nupic/ntypes/ArrayBase.cpp

```diff
@@ -181,27 +181,28 @@ void *ArrayBase::getBuffer() {
 const void *ArrayBase::getBuffer() const {
   if (has_buffer()) {
     if (buffer_ != nullptr && type_ == NTA_BasicType_SDR) {
-      return getSDR()->getDense().data();
+      return getSDR().getDense().data();
     }
     return buffer_.get();
   }
   return nullptr;
 }
 
-SDR *ArrayBase::getSDR() {
+SDR& ArrayBase::getSDR() {
   NTA_CHECK(type_ == NTA_BasicType_SDR) << "Does not contain an SDR object";
-  SDR *sdr = (SDR *)buffer_.get();
-  if (sdr == nullptr)
-    NTA_THROW << "getSDR: SDR pointer is null";
-  sdr->setDense(sdr->getDense()); // cleanup cache
+  if (buffer_ == nullptr)
+    allocateBuffer({ 0 });  // Create an empty SDR object.
+  SDR& sdr = *((SDR *)buffer_.get());
```

Comment:
what type is buffer_.get()?

Reply:
Would there be a nice way to do this?

Reply:
> what type is buffer_.get()?

It is a char pointer.

>Would there be a nice way to do this?

Well...not really.  It is not type-safe no matter how I do it since the pointer could actually be one of several types, one of which is SDR*.   I would like to use something like std:any but that is not available until C++17.   It is the same problem we have with getBuffer( ).

Reply:
>  I would like to use something like std:any but that is not available until C++17. It is the same problem we have with getBuffer( ).


do you want to write #IFDEF code for (not)c++17? When we hard-switch, we can use it easily.

Reply:
> do you want to write #IFDEF code for (not)c++17? When we hard-switch, we can use it easily.

Well, perhaps but getBuffer( ) is used in a lot of places and it needs the same fix.  Perhaps later.  I don't see that as a high priority.